llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -248,34 +249,36 @@ void dumpGraph(int NInstr, const std::vector<NodeInfo> &Info,
   dbgs() << "}\n";
 }
 
-int PostPipeliner::mostUrgent() {
-  assert(FirstUnscheduled < NInstr);
+int PostPipeliner::mostUrgent(PostPipelinerStrategy &Strategy) {
+  assert(FirstUnscheduled <= LastUnscheduled);
   while (Info[FirstUnscheduled].Scheduled) {
     FirstUnscheduled++;
   }
-  assert(FirstUnscheduled < NInstr);
+  while (Info[LastUnscheduled].Scheduled) {
+    LastUnscheduled--;
+  }
+  assert(FirstUnscheduled <= LastUnscheduled);
 
-  // 'Latest' accounts for the critical path of the linear schedule
-  auto Better = [this](int N, int Ref) {
-    if (Info[N].Latest < Info[Ref].Latest) {
-      return true;
+  auto NotScheduled = [&](const auto &Dep) {
+    auto *SU = Dep.getSUnit();
+    if (SU->isBoundaryNode()) {
+      return false;
     }
-
-    return false;
+    int N = SU->NodeNum;
+    return N < NInstr && !Info[N].Scheduled;
   };
 
   int Best = -1;
   LLVM_DEBUG(dbgs() << "Available:");
-  for (int K = FirstUnscheduled; K < NInstr; K++) {
+  for (int K = FirstUnscheduled; K <= LastUnscheduled; K++) {
```

Comment:
Super-Nit: Use a `filter_range` to only iterate on the non-scheduled nodes. Then we do not have to verify `Info[K].Scheduled` inside the loop.

Reply:
I think a filter range is just a way to verify a condition in a loop. I reserve its use to write a generic iteration template, not to define a single instance.